pkg/networkpath/traceroute/tcp/utils.go

```diff
@@ -0,0 +1,423 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package tcp
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"go.uber.org/multierr"
+	"golang.org/x/net/ipv4"
+)
+
+const (
+	// URG is the urgent TCP flag
+	URG = 1 << 5
+	// ACK is the acknowledge TCP flag
+	ACK = 1 << 4
+	// PSH is the push TCP flag
+	PSH = 1 << 3
+	// RST is the reset TCP flag
+	RST = 1 << 2
+	// SYN is the synchronization TCP flag
+	SYN = 1 << 1
+	// FIN is the final TCP flag
+	FIN = 1 << 0
+)
+
+type (
+	// canceledError is sent when a listener
+	// is canceled
+	canceledError string
+
+	// icmpResponse encapsulates the data from
+	// an ICMP response packet needed for matching
+	icmpResponse struct {
+		SrcIP        net.IP
+		DstIP        net.IP
+		TypeCode     layers.ICMPv4TypeCode
+		InnerSrcIP   net.IP
+		InnerDstIP   net.IP
+		InnerSrcPort uint16
+		InnerDstPort uint16
+		InnerSeqNum  uint32
+	}
+
+	// tcpResponse encapsulates the data from a
+	// TCP response needed for matching
+	tcpResponse struct {
+		SrcIP       net.IP
+		DstIP       net.IP
+		TCPResponse *layers.TCP
+	}
+)
+
+func localAddrForHost(destIP net.IP, destPort uint16) (*net.UDPAddr, error) {
+	// this is a quick way to get the local address for connecting to the host
+	// using UDP as the network type to avoid actually creating a connection to
+	// the host, just get the OS to give us a local IP and local ephemeral port
+	conn, err := net.Dial("udp4", net.JoinHostPort(destIP.String(), strconv.Itoa(int(destPort))))
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
```

Comment:
Wouldn't this free up the local port?

Reply:
It does, we don't really care about binding to the port, I'm just using this to get a random ephemeral port number to use from the OS to send on. Since we have a raw connection we can set whatever source information we want.